### PR TITLE
Add kustomizations and tooling to create 2.x manifests

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:
- Adds Kustomize patches to change the image in deploy manifest bases to 2.x and use its readiness endpoint
- Adds a script (and make target) to copy the 2.x kubebuilder-generated manifests and existing 1.x deploy manifests into tempdir and kustomize them into single-file manifests under `deploy/single-v2`
- Move existing manifests to `deploy/single-v1` and link `deploy/single` to `deploy/single-v1`. We should flip this to v2 once v2 is GA.
- Update manifest verify script to use new locations and check v2 manifests.

**Which issue this PR fixes**: Fixes https://github.com/Kong/kubernetes-ingress-controller/issues/1256

**Special notes for your reviewer**:
- This PR includes the generated manifests, and you probably don't want to review them line by line. `kubectl apply -f deploy/single-v2/all-in-one-dbless.yaml` and so on in a test cluster is a more effective review strategy.
- Main files to review are [hack/verify-manifests.sh](https://github.com/Kong/kubernetes-ingress-controller/pull/1563/files?file-filters%5B%5D=#diff-b01a20d252dbad1988a8985e3bd908620249f7804c55598e7f2c0a73bef48516), [railgun/config/base/kong-ingress-2x.yaml](https://github.com/Kong/kubernetes-ingress-controller/pull/1563/files?file-filters%5B%5D=#diff-069887bb9c8560e72c8384561f5042563e6ccfeba34347493d29c456f2a777e3), [railgun/config/base/kustomization.yaml ](https://github.com/Kong/kubernetes-ingress-controller/pull/1563/files?file-filters%5B%5D=.yaml#diff-81a9e4fe5b30ea74334713ea2ac4af488353f91ec928c74c4239d5b8148467e3), and [railgun/hack/deploy/build-single-manifests.sh](https://github.com/Kong/kubernetes-ingress-controller/pull/1563/files?file-filters%5B%5D=#diff-cd8e3747bbc33b415baa275b33d9b202145ffc136c3c5ab2f511d9f658126eb8). The rest are largely repetitive and/or generated manifests composed from existing content (e.g. there's nothing new about the CRDs or RBAC resources, this just composes them into new files).
- There are two open TODOs:
  - [x] Fix the role bindings. Does anyone (probably @shaneutt) recall if [the binding](https://github.com/Kong/kubernetes-ingress-controller/blob/a8cd5833b4ebeca1cbe458d24d946c60c175039e/railgun/config/rbac/role_binding.yaml) was written by hand or generated? I think  it was written by hand and I can just modify it directly and remove `tmp-bindings.yaml`. If it was generated, however, I need to fix the source, and I don't know what that is.
  - [ ] This doesn't deal with the [auth proxy role](https://book.kubebuilder.io/reference/metrics.html#protecting-the-metrics). Are we okay taking that as a followup item?

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Comment by rainest** [comment]
Draft version of this attempts to more or less mimic the 1.x hack script, using the existing 1.x kustomize bases with some additional patches for 2.x. 2.x bases symlink to the 1.x bases when they don't need changes.

Open questions to refine the draft into a ready PR:
- Does the file organization proposed here make sense? This is sort of an arbitrary organic layout I threw together to make it work that sorta tries to follow the existing base layout. I don't know how much we like/don't like the existing structure under `deploy/` and `railgun/config/` and/or whether it makes intuitive sense, and don't have much opinion about it myself: it's there and I think I understand it, but I am not a particularly organized person and will thrive in whatever chaotic mess I find :upside_down_face: 
- Does this approach to maintaining a 1.x/2.x split seem sound? From initial basic testing (run `make allinones` and deploy the resulting 2.x single manifests to KIND, see if they come up without errors), the existing 1.x deployment files are fine with minimal edits to change the version, and the main 2.x stuff is to compose the new generated CRDs and RBAC resources alongside the deployments.
- What's a better name for the make target? `make allinones` is my extremely eNgLiSh "make is a transitive verb whose object must desribe its output, and its output is 'allinones', a neonoun for 'all in ones', the combined YAML manifests including Deployment, CRD, RBAC, etc. There's probably a more sensible name for this, but my brain is stuck on "well, that's what it outputs" at the moment.

File: railgun/config/base/kustomization.yaml
Comment (ccfishk) [inline railgun/config/base/kustomization.yaml:1]: Do you plan to consider maximize usage of files under deploy/manifests, deploy/single to remove duplicates within this PR ?

**Review by ccfishk — CHANGES_REQUESTED** [review]
Kindly help to remove duplicates

File: railgun/config/base/kustomization.yaml
Comment (rainest) [reply railgun/config/base/kustomization.yaml:1]: Yes, hence why, for example, ` railgun/config/base/kong-ingress-dbless.yaml` is just a symlink to ` ../../../deploy/manifests/base/kong-ingress-dbless.yaml `. That's a bit clearer on the filesystem than it is in Github's UI:

```
$ ls railgun/config/base -l
total 16
-rw-r--r-- 1 traines traines  298 Jul 21 10:30 kong-ingress-2x.yaml
lrwxrwxrwx 1 traines traines   55 Jul 15 14:54 kong-ingress-dbless.yaml -> ../../../deploy/manifests/base/kong-ingress-dbless.yaml
-rw-r--r-- 1 traines traines  316 Jul 16 11:23 kustomization.yaml
lrwxrwxrwx 1 traines traines   45 Jul 15 14:54 namespace.yaml -> ../../../deploy/manifests/base/namespace.yaml
-rw-r--r-- 1 traines traines   92 Jul 15 15:35 serviceaccount.yaml
lrwxrwxrwx 1 traines traines   43 Jul 15 14:53 service.yaml -> ../../../deploy/manifests/base/service.yaml
-rw-r--r-- 1 traines traines 1089 Jul 16 11:03 tmp-bindings.yaml
lrwxrwxrwx 1 traines traines   54 Jul 15 14:54 validation-service.yaml -> ../../../deploy/manifests/base/validation-

$ ls railgun/config/variants/postgres -l
total 8
lrwxrwxrwx 1 traines traines  64 Jul 16 11:35 kong-ingress-postgres.yaml -> ../../../../deploy/manifests/postgres/kong-ingress-postgres.yaml
-rw-r--r-- 1 traines traines 110 Jul 16 11:25 kustomization.yaml
lrwxrwxrwx 1 traines traines  52 Jul 16 11:36 migration.yaml -> ../../../../deploy/manifests/postgres/migration.yaml
lrwxrwxrwx 1 traines traines  51 Jul 16 11:36 postgres.yaml -> ../../../../deploy/manifests/postgres/postgres.yaml
```
The new content for each of the 2.x variants is kustomization.yamls (which need to be unique), patches, and the ServiceAccount (because 1.x had it included with the 1.x RBAC roles, which we don't want in 2.x).

kustomization.yaml could point to the 1.x templates directly instead, but IMO the symlink version is more readable.

File: railgun/config/base/tmp-bindings.yaml
Comment (ccfishk) [inline railgun/config/base/tmp-bindings.yaml:1]: can we re-use previous role.yaml ? I think we already fix adding the udp api group and operations in single one.yaml .

File: railgun/config/crd/kustomization.yaml
Comment (ccfishk) [inline railgun/config/crd/kustomization.yaml:7]: I believe we have those files under deploy/manifests, deploy/single

File: railgun/config/variants/enterprise/kustomization.yaml
Comment (ccfishk) [inline railgun/config/variants/enterprise/kustomization.yaml:8]: please do not hardcode tag, otherwise, we need add duplicate yaml each round.

**Review by ccfishk — CHANGES_REQUESTED** [review]
a few comments

File: railgun/config/crd/kustomization.yaml
Comment (rainest) [reply railgun/config/crd/kustomization.yaml:7]: Can you clarify? What's the requested change or question here?

File: railgun/config/variants/enterprise/kustomization.yaml
Comment (rainest) [reply railgun/config/variants/enterprise/kustomization.yaml:8]: This is following the convention we use for 1.x, where both the Enterprise variants override their image in kustomization.yaml:
https://github.com/Kong/kubernetes-ingress-controller/blob/1.3.1/deploy/manifests/enterprise/kustomization.yaml
https://github.com/Kong/kubernetes-ingress-controller/blob/1.3.1/deploy/manifests/enterprise-k8s/kustomization.yaml

That is an artifact from an earlier era where database-backed and DB-less Enterprise clusters did actually use different images. We probably can now look at making them cascade, where only only the Enterprise DB-less kustomization.yaml overrides the image and the Enterprise Postgres variant builds off it, but it'd require some refactoring to do correctly, since we currently build Enterprise Postgres off of OSS Postgres, not Enterprise DB-less.

Updating the images in each is simple grep and replace, so my vote is to defer that refactor; it should be separate from the work here.

File: railgun/config/crd/kustomization.yaml
Comment (ccfishk) [reply railgun/config/crd/kustomization.yaml:7]: Can you please re-use this yaml https://github.com/Kong/kubernetes-ingress-controller/blob/main/deploy/manifests/base/custom-types.yaml#L5 or remove duplicates ?

File: deploy/single
Comment (ccfishk) [inline deploy/single:1]: Do you mind explain where does the following single-v2 come from ?

File: hack/build-single-manifests.sh
Comment (ccfishk) [inline hack/build-single-manifests.sh:13]: looks like you have empty single-v1, but some data under single-v2 ?

File: railgun/config/base/kong-ingress-2x.yaml
Comment (ccfishk) [inline railgun/config/base/kong-ingress-2x.yaml:11]: should not this be kic beta release ?

File: railgun/config/base/tmp-bindings.yaml
Comment (ccfishk) [reply railgun/config/base/tmp-bindings.yaml:1]: Are you saying following same file generated using tools ? 
https://github.com/Kong/kubernetes-ingress-controller/blob/350d0da8062392c7570c9f7f40e3558a5885a772/deploy/manifests/base/rbac.yaml

I believe they are manually created as you did.

File: railgun/config/variants/enterprise/kustomization.yaml
Comment (ccfishk) [inline railgun/config/variants/enterprise/kustomization.yaml:1]: same as 
railgun/config/variants/enterprise-postgres/kustomization.yaml

Are you sure about so many duplicates ?

**Comment by ccfishk** [comment]
Kindly split the files and remove duplicates.

**Review by ccfishk — COMMENTED** [review]
added comments

File: railgun/config/base/tmp-bindings.yaml
Comment (rainest) [reply railgun/config/base/tmp-bindings.yaml:1]: `deploy/manifests/base/rbac.yaml` was written manually. `railgun/config/rbac/role.yaml` was generated from type annotations.

**Review by ccfishk — COMMENTED** [review]
commented.

File: railgun/config/crd/kustomization.yaml
Comment (rainest) [reply railgun/config/crd/kustomization.yaml:7]: Similar to the RBAC resources, 2.x CRDs are generated by kubebuilder. We specifically do not want to reuse the 1.x CRDs.

File: deploy/single
Comment (rainest) [reply deploy/single:1]: Is this a question about the layout or the contents? Manifests under single-v1 are generated by `hack/build-single-manifests.sh`. Manifests under single-v2 are generated by `railgun/hack/deploy/build-single-manifests.sh`.

The layout is from my understanding of Michal's request this morning, that we place both the v1 and v2 manifests in the same location, which I thought was reasonable. The addition of the `single -> single-v1` symlink was to ensure that existing references elsewhere do not break (we have a number of bit.ly links in documentation that point to the existing locations under `single`).

The Github UI display for symlinks is weird--it sorta looks like this is instead a file with a single line of content--but this is indeed a symlink.

File: hack/build-single-manifests.sh
Comment (rainest) [reply hack/build-single-manifests.sh:13]: Both of them do have content:

```
$ git status
On branch feat/rg-deploy-tooling
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	cli/ingress-controller/ingress-controller
	railgun/hack/generators/controllers/networking/networking
	rbac.md

nothing added to commit but untracked files present (use "git add" to track)

$ git rev-parse HEAD
843cfc61343ad92db3a535d5fed617c764238857

$ ls deploy/single-v1 
all-in-one-dbless-k4k8s-enterprise.yaml  all-in-one-dbless.yaml  all-in-one-postgres-enterprise.yaml  all-in-one-postgres.yaml

$ ls deploy/single-v2
all-in-one-dbless.yaml	all-in-one-enterprise-dbless.yaml  all-in-one-enterprise-postgres.yaml	all-in-one-postgres.yaml
```

File: railgun/config/base/kong-ingress-2x.yaml
Comment (rainest) [reply railgun/config/base/kong-ingress-2x.yaml:11]: Not until that image actually exists, no. Updating these with the upcoming release version is part of our release process; when we're preparing to release 2.0.0-beta.1 we'll change it.

File: railgun/config/base/tmp-bindings.yaml
Comment (ccfishk) [reply railgun/config/base/tmp-bindings.yaml:1]: Right. 
Assuming helm chart installation will consume the new one railgun/role.yaml 
The real challenge right now is how to proceed to remove duplicates.

File: railgun/config/variants/enterprise/kustomization.yaml
Comment (rainest) [reply railgun/config/variants/enterprise/kustomization.yaml:1]: They are different:

```
$ diff railgun/config/variants/enterprise/kustomization.yaml railgun/config/variants/enterprise-postgres/kustomization.yaml  
2c2,3                      
< - ../../base/
---
> - ../postgres
> - enterprise-resources.yaml
```
The new 2.x kustomization.yamls mirror the 1.x kustomization.yamls. Each provides an OSS DB-less, OSS Postgres, Enterprise DB-less, and Enterprise Postgres set of manifests.

File: railgun/config/base/tmp-bindings.yaml
Comment (rainest) [reply railgun/config/base/tmp-bindings.yaml]: After searching around a bit more, I think we're good to remove the temp patch in favor of modifying the source manifest in ff2ca18309138e20688e841900dff71a2cdbffaf

The source hasn't changed since we created the railgun directory. Were it auto-generated, it'd likely have _some_ change since then.

File: deploy/single
Comment (rainest) [reply deploy/single:1]: Resolved during a call. Github symlink display is just confusing :shrug: :shrimp:

File: railgun/config/base/kong-ingress-2x.yaml
Comment (shaneutt) [reply railgun/config/base/kong-ingress-2x.yaml:11]: NOTE: if we have several places where we need to manually update versions we should note that in the RELEASE.md notes, and create follow up Github issues to later automate that.

File: railgun/config/variants/enterprise-postgres/kustomization.yaml
Comment (shaneutt) [inline railgun/config/variants/enterprise-postgres/kustomization.yaml:9]: Similar to my above comment, since we're using such a specific version and security issues might get missed we should add a follow up Github issue to automate updates of this tag.

File: railgun/hack/deploy/build-single-manifests.sh
Comment (shaneutt) [inline railgun/hack/deploy/build-single-manifests.sh:21]: It looks like we need to resolve this TODO before merging.

**Review by shaneutt — APPROVED** [review]
Looks good, I have no hard blocks :+1: 

I do however have a few comments about follow ups for some of these hardcoded versions, and to make sure we're following up on the TODO item from the script, once those are complete and other people's comments are complete :heavy_check_mark:

**Review by ccfishk — APPROVED** [review]
LGTM after removing the tmp-binding.yaml. Thanks for working hard removing duplicates, I know it is challenge, but we'll get there.

File: railgun/hack/deploy/build-single-manifests.sh
Comment (rainest) [reply railgun/hack/deploy/build-single-manifests.sh:21]: It was something I wasn't familiar with as much and wasn't immediately necessary (since it's only used for Prometheus). It also requires some environment-specific configuration (you must bind to your Prometheus user; we don't know who that is in advance) that we can't add to the generated manifests. https://github.com/Kong/kubernetes-ingress-controller/issues/1579

**Pull request merged**
